### PR TITLE
Benchmarks CompletableFuture, reactor.Mono, rxjava.Single #514

### PR DESCRIPTION
**Pull request by fkowal**
Here is a first batch of changes, waiting for comments.

Take a look at usages of CompletableFuture.completedFuture(n), the behaviour is not consistent with other tests.
The signature is (a: A) where as other frameworks have (a: => A), but i am not quite sure what would be a more suitable choice.

### CODE REVIEW

File: benchmarks/src/main/scala/scalaz/zio/ArrayFillBenchmark.scala [hunk 1]
```diff
@@ -36,6 +36,25 @@ class ArrayFillBenchmarks {
       } yield ()
     )
   }
+
+  @Benchmark
+  def monoArrayFill() = {
+    import reactor.core.publisher.Mono
+
+    def arrayFill(array: Array[Int])(i: Int): Mono[Unit] =
+      if (i >= array.length) Mono.just(())
```
Comment (jdegoes) [inline benchmarks/src/main/scala/scalaz/zio/ArrayFillBenchmark.scala]: You should use the lazy version of `just`, whatever that is.

File: benchmarks/src/main/scala/scalaz/zio/IODeepAttemptBenchmark.scala [hunk 2]
```diff
@@ -40,6 +41,68 @@ class IODeepAttemptBenchmark {
...
+        val f = new CompletableFuture[BigInt]()
+        f.completeExceptionally(new Exception("Oh noes!"))
+        f
+      } else if (n == halfway) {
+        descent(n + 1).exceptionally(_ => 50)
+      } else {
+        descent(n + 1).thenApply(_ + n)
+      }
+
+    descent(0).get()
```
Comment (jdegoes) [inline benchmarks/src/main/scala/scalaz/zio/IODeepAttemptBenchmark.scala]: This does not look lazy to me. Is there some way you can fix that?

File: benchmarks/src/main/scala/scalaz/zio/IODeepAttemptBenchmark.scala [hunk 3]
```diff
@@ -40,6 +41,68 @@ class IODeepAttemptBenchmark {
...
+  def twitterDeepAttempt(): BigInt = {
+    import com.twitter.util.{ Await, Future }
+
+    def descent(n: Int): Future[BigInt] =
+      if (n == depth)
+        Future.exception(new Error("Oh noes!"))
+      else if (n == halfway)
+        descent(n + 1).handle { case _ => 50 } else descent(n + 1).map(_ + n)
+
+    Await.result(descent(0))
```
Comment (jdegoes) [inline benchmarks/src/main/scala/scalaz/zio/IODeepAttemptBenchmark.scala]: Ditto.

File: benchmarks/src/main/scala/scalaz/zio/ArrayFillBenchmark.scala (hunk 1 above)
Comment (jdegoes) [reply benchmarks/src/main/scala/scalaz/zio/ArrayFillBenchmark.scala]: An approximation would be: `flatten(just(just(()))`, perhaps?

Comment (fkowal) [reply benchmarks/src/main/scala/scalaz/zio/ArrayFillBenchmark.scala]: .fromCallable(() => ()) ?

File: benchmarks/src/main/scala/scalaz/zio/IODeepAttemptBenchmark.scala (hunk 2 above)
Comment (fkowal) [reply benchmarks/src/main/scala/scalaz/zio/IODeepAttemptBenchmark.scala]: both CompletableFuture, and twitter.Future are eagerly evaluated
and are a reflection of scala Future benchmarks implementation

i don't really see a way to change this behaviour

**Pull request merged**